data_model/src/error.rs

```diff
@@ -0,0 +1,243 @@
+    /// Failed to find specified [`Parameter`] variant.
+    #[error("Failed to find specified parameter variant: `{0}`")]
+    Parameter(Parameter),
+}
+
+/// Generic structure used to represent a mismatch
+#[derive(Debug, Clone, PartialEq, Eq, Error, Decode, Encode, IntoSchema)]
+#[error("Expected {expected:?}, actual {actual:?}")]
+pub struct Mismatch<T> {
+    /// The value that is needed for normal execution
+    pub expected: T,
+    /// The value that caused the error
+    pub actual: T,
+}
+
+/// Type error
+#[derive(Debug, Clone, Error, PartialEq, Eq)]
+#[allow(variant_size_differences)] // False-positive
+pub enum TypeError {
+    /// Asset type assertion error
+    #[error("Asset Ids correspond to assets with different underlying types. {0}")]
+    AssetValueType(#[from] Mismatch<AssetValueType>),
+    /// Parameter type assertion error
+    #[error("Value passed to the parameter doesn't have the right type. {0}")]
+    ParameterValueType(#[from] Mismatch<Value>),
+    /// Asset Id mismatch
+    #[error("AssetDefinition Ids don't match. {0}")]
+    AssetDefinitionId(#[from] Box<Mismatch<<AssetDefinition as Identifiable>::Id>>),
+}
+
+/// Math error, which occurs during instruction execution
+#[derive(Debug, Clone, Error, Copy, PartialEq, Eq)]
+pub enum MathError {
+    /// Overflow error inside instruction
+    #[error("Overflow occurred.")]
+    Overflow,
+    /// Not enough quantity
+    #[error("Not enough quantity to transfer/burn.")]
+    NotEnoughQuantity,
+    /// Divide by zero
+    #[error("Divide by zero")]
+    DivideByZero,
+    /// Negative Value encountered
+    #[error("Negative value encountered")]
+    NegativeValue,
+    /// Domain violation
+    #[error("Domain violation")]
+    DomainViolation,
+    /// Unknown error. No actual function should ever return this if possible.
+    #[error("Unknown error")]
+    Unknown,
+    /// Encountered incompatible type of arguments
+    #[error("Binary operation does not support provided combination of arguments ({0}, {1})")]
+    BinaryOpIncompatibleNumericValueTypes(NumericValue, NumericValue),
+}
+
+impl From<FixedPointOperationError> for IsiError {
+    fn from(err: FixedPointOperationError) -> Self {
+        match err {
+            FixedPointOperationError::NegativeValue(_) => Self::Math(MathError::NegativeValue),
+            FixedPointOperationError::Conversion(e) => {
+                Self::Conversion(format!("Mathematical conversion failed. {e}"))
+            }
+            FixedPointOperationError::Overflow => MathError::Overflow.into(),
+            FixedPointOperationError::DivideByZero => MathError::DivideByZero.into(),
+            FixedPointOperationError::DomainViolation => MathError::DomainViolation.into(),
+            FixedPointOperationError::Arithmetic => MathError::Unknown.into(),
+        }
+    }
+}
+
+impl From<FindError> for IsiError {
+    fn from(err: FindError) -> Self {
+        Self::Find(Box::new(err))
+    }
+}
```

Comment:
We could consider replacing `thiserror` with `derive_more`, given that this pattern would be covered with `derive_more::From`, and we can derive `Display` independently from `error`.

Reply:
Created #3267.